### PR TITLE
Fix for issue #452

### LINKED ISSUE #452: Resource IDs with spaces in them cause an invalid location URI to be generated
Some end systems use resource IDs which have spaces in the value. This causes a validation error when SCIM-SDK is generating the location metadata for a resource as the resource ID is not currently being URL escaped.

I have modified a version of the springboot example which demonstrates the issue here: https://github.com/mberning/SCIM-SDK/compare/master...bad-location-uri-example

Example: A Group with a resource ID of "super admin"

Request:
```bash
curl "http://localhost:8080/scim/v2/Groups"
```

Response:
```json
{
  "detail": "An internal error has occurred.",
  "schemas": [
    "urn:ietf:params:scim:api:messages:2.0:Error"
  ],
  "status": 500
}
```

Stacktrace:
```
24 Mar 2023 16:20:24,766 [http-nio-8080-exec-2] ERROR de.captaingoldfish.scim.sdk.common.response.ErrorResponse (ErrorResponse.java:63) - Attribute 'urn:ietf:params:scim:schemas:core:2.0:Meta:meta.location' is a referenceType and must apply to one of the following types '[URI]' but value is 'http://localhost:8080/scim/v2/Groups/super admin'
de.captaingoldfish.scim.sdk.common.exceptions.InternalServerException: Attribute 'urn:ietf:params:scim:schemas:core:2.0:Meta:meta.location' is a referenceType and must apply to one of the following types '[URI]' but value is 'http://localhost:8080/scim/v2/Groups/super admin'
	at de.captaingoldfish.scim.sdk.server.endpoints.ResourceEndpointHandler.listResources(ResourceEndpointHandler.java:629)
	at de.captaingoldfish.scim.sdk.server.endpoints.ResourceEndpoint.resolveRequest(ResourceEndpoint.java:255)
	at de.captaingoldfish.scim.sdk.server.endpoints.ResourceEndpoint.handleRequest(ResourceEndpoint.java:190)
	at de.captaingoldfish.scim.sdk.server.endpoints.ResourceEndpoint.handleRequest(ResourceEndpoint.java:77)
	at de.captaingoldfish.scim.sdk.springboot.sample.controller.ScimController.handleScimRequest(ScimController.java:68)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:207)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:152)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:117)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:884)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:797)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1080)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:973)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1011)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:903)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:705)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:885)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:814)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:223)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:158)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:185)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:158)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:185)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:158)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:177)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:542)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:119)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:357)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:400)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:859)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1734)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:1589)
Caused by: de.captaingoldfish.scim.sdk.server.schemas.exceptions.AttributeValidationException: Attribute 'urn:ietf:params:scim:schemas:core:2.0:Meta:meta.location' is a referenceType and must apply to one of the following types '[URI]' but value is 'http://localhost:8080/scim/v2/Groups/super admin'
	at de.captaingoldfish.scim.sdk.server.schemas.validation.SimpleAttributeValidator.validateValueNodeWithReferenceTypes(SimpleAttributeValidator.java:206)
	at de.captaingoldfish.scim.sdk.server.schemas.validation.SimpleAttributeValidator.parseNodeType(SimpleAttributeValidator.java:125)
	at de.captaingoldfish.scim.sdk.server.schemas.validation.ValidationSelector.validateNode(ValidationSelector.java:83)
	at de.captaingoldfish.scim.sdk.server.schemas.validation.ComplexAttributeValidator.parseNodeType(ComplexAttributeValidator.java:50)
	at de.captaingoldfish.scim.sdk.server.schemas.validation.ValidationSelector.validateNode(ValidationSelector.java:76)
	at de.captaingoldfish.scim.sdk.server.schemas.validation.ResponseAttributeValidator.validateAttribute(ResponseAttributeValidator.java:245)
	at de.captaingoldfish.scim.sdk.server.schemas.validation.ResponseSchemaValidator.validateAttribute(ResponseSchemaValidator.java:63)
	at de.captaingoldfish.scim.sdk.server.schemas.validation.AbstractSchemaValidator.validateDocument(AbstractSchemaValidator.java:128)
	at de.captaingoldfish.scim.sdk.server.schemas.validation.ResponseResourceValidator.validateDocument(ResponseResourceValidator.java:48)
	at de.captaingoldfish.scim.sdk.server.endpoints.ResourceEndpointHandler.listResources(ResourceEndpointHandler.java:616)
	... 44 more
```

I will submit a PR which resolves this issue and reference this issue in the commit messages.

### PR DESCRIPTION
**Pull request by mberning**
This change fixes #452 by URL encoding resource identifiers.

### CODE REVIEW

File: scim-sdk-common/src/main/java/de/captaingoldfish/scim/sdk/common/utils/EncodingUtils.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+public class EncodingUtils
+{
+
+  /**
+   * This method takes the input string and URL Encodes it using java.net.URLEncoder
+   *
+   * @param value The value to be URL encoded.
+   * @return A URL encoded version of the input value
+   */
+  public static String urlEncode(String value)
```
Comment (Captain-P-Goldfish) [inline scim-sdk-common/src/main/java/de/captaingoldfish/scim/sdk/common/utils/EncodingUtils.java:21]: could you move this into the class `UtilityMethods`?
I would prefer to have fewer than more classes of this kind :-)

**Review by Captain-P-Goldfish — COMMENTED** [review]
I also found another problem:
The test-method in `ResourceEndpointTest#testGetResource()`. Try to modify the testcase like this:

```
final String userName = "chuck_norris";
final String id = EncodingUtils.urlEncode("$chuck- norris!");
```

You will see that a problem will arise that might cause trouble under certain circumstances. This should be fixed next.

File: scim-sdk-common/src/main/java/de/captaingoldfish/scim/sdk/common/utils/EncodingUtils.java (hunk 1 above)
Comment (mberning) [reply scim-sdk-common/src/main/java/de/captaingoldfish/scim/sdk/common/utils/EncodingUtils.java:21]: Sure, no problem, I will make the change.

**Comment by mberning** [comment]
1. Fixed the areas where meta location provided by the developer implementation was being overridden and added unit tests
2. I looked at moving the encoding utility methods to the `UtilityMethods` class, but it is not currently available in `scim-sdk-server`. Not sure how you would like to proceed.
3. Fixing the bug where location was not being respected caused `ResourceTypeEndpointTest` to start failing. I am a bit concerned that it was depending on the old behavior. I started looking in to how the meta location is being set for the resource types, but wanted to get your input before proceeding.

**Comment by Captain-P-Goldfish** [comment]
Seems like I broke CircleCi integration. I tried to repair it now:

1. There are still formatting causing the maven-build to error
2. 5 test error but I would be willing to fix them myself. You will have to modify  the schema files in order to fix these tests
3. The methods from the class `EncodingUtils` were not moved as discussed.

**Comment by mberning** [comment]
> Seems like I broke CircleCi integration. I tried to repair it now:
> 
>     1. There are still formatting causing the maven-build to error. You can fix the formatting with `mvn formatter:format`
> 
>     2. 5 test errors but I would be willing to fix them myself. You will have to modify  the schema files in order to fix these tests
> 
>     3. The methods from the class `EncodingUtils` were not moved as discussed.

I pushed commits to resolve point 1 and 2. On point 3 I looked at moving the encoding utility methods to the `UtilityMethods` class, but it is not currently available in `scim-sdk-server`. Not sure how you would like to proceed.

**Comment by Captain-P-Goldfish** [comment]
ignore point 3 :-)
I was somehow sure it was in the common module. my mistake

**Comment by Captain-P-Goldfish** [comment]
Could you simply delete the location attributes in the schemas? Otherwise these schemas would be returned with erroneous locations in their location attribute from the /ResourceTypes endpoint.
Afterwards try to build with maven. You will notice that the build fails because I have entered checksums in the scim-sdk-common/pom.xml file to notice unintentional changes on these files.

**Comment by mberning** [comment]
> Could you simply delete the location attributes in the schemas? Otherwise these schemas would be returned with erroneous locations in their location attribute from the /ResourceTypes endpoint. Afterwards try to build with maven. You will notice that the build fails because I have entered checksums in the scim-sdk-common/pom.xml file to notice unintentional changes on these files.

I gave this a try and ran into an issue with https://github.com/Captain-P-Goldfish/SCIM-SDK/blob/master/scim-sdk-server/src/main/java/de/captaingoldfish/scim/sdk/server/schemas/ResourceType.java#L117

Seems that is is generating a location when none is provided, but is not aware of the current base URI.

**Comment by mberning** [comment]
I pushed a fix and updated the checksums

**Pull request merged**